### PR TITLE
cov: add tests for ignite.contrib.engines

### PR DESCRIPTION
**Pull request by ydcjeff**
Part of #1790 

Description: Add tests for `ignite.contrib.engines`

Check list:

- [x] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/contrib/engines/test_common.py [hunk 1]
```diff
@@ -165,7 +188,7 @@ def test_assert_setup_common_training_handlers_wrong_train_sampler(distributed_c
 
 
 def test_setup_common_training_handlers(dirname, capsys):
-
+    print(dirname)
```
Comment (sdesrozis) [inline tests/ignite/contrib/engines/test_common.py:191]: Please remove this line

File: ignite/handlers/checkpoint.py [hunk 2]
```diff
@@ -708,7 +709,7 @@ def _save_func(self, checkpoint: Mapping, path: str, func: Callable, rank: int =
             else:
                 if tmp is not None:
                     tmp.close()
-                    os.rename(tmp.name, path)
+                    shutil.move(tmp.name, path)
```
Comment (sdesrozis) [inline ignite/handlers/checkpoint.py:712]: Why this in the PR?

Comment (ydcjeff) [reply ignite/handlers/checkpoint.py:712]: Windows has `FileExistError` with `os.rename`, so i changed to `shutil.move`.

File: tests/ignite/contrib/engines/test_common.py (hunk 1 above)
Comment (ydcjeff) [inline tests/ignite/contrib/engines/test_common.py:191]:
```suggestion
```

File: ignite/handlers/checkpoint.py (hunk 2 above)
Comment (sdesrozis) [reply ignite/handlers/checkpoint.py:712]: It means that was not functional on macOS and windows ? Did you state that introducing a new test ?

Comment (ydcjeff) [reply ignite/handlers/checkpoint.py:712]: I added [these extra tests](https://github.com/pytorch/ignite/pull/1808/files#diff-bd52578edbfac3a9213886eb448a36cf7c13ad3c95110279ccf1d76c9b4a35a7R200-R217) to cover missing parts and `test_setup_common_training_handlers` test passed on mac and linux, failed on windows resulting an error of `FileExistError` with `os.rename`.
So I changed to `shutil.move`

**Review by sdesrozis — APPROVED** [review]
@ydcjeff Thank you!

**Comment by sdesrozis** [comment]
@ydcjeff Windows tests are ko. Could you have a look ?

**Comment by ydcjeff** [comment]
@sdesrozis windows tests are failing on `test_average_precision`
```py
=========================== short test summary info ===========================
FAILED tests/ignite/contrib/metrics/test_average_precision.py::test_binary_input_N
FAILED tests/ignite/contrib/metrics/test_average_precision.py::test_multilabel_input_N
```
Error is metric is computing nan
```py
>       assert average_precision_score(np_y, np_y_pred) == pytest.approx(res)
E       assert nan == nan � ???
E         +nan
E         -nan � ???
```

**Comment by vfdev-5** [comment]
> @sdesrozis windows tests are failing on `test_average_precision`
> 
> ```python
> =========================== short test summary info ===========================
> FAILED tests/ignite/contrib/metrics/test_average_precision.py::test_binary_input_N
> FAILED tests/ignite/contrib/metrics/test_average_precision.py::test_multilabel_input_N
> ```
> 
> Error is metric is computing nan
> 
> ```python
> >       assert average_precision_score(np_y, np_y_pred) == pytest.approx(res)
> E       assert nan == nan � ???
> E         +nan
> E         -nan � ???
> ```

@KickItLikeShika have you seen this ?

**Comment by KickItLikeShika** [comment]
@vfdev-5 This never happened before..

**Comment by KickItLikeShika** [comment]
@vfdev-5 i don't think it's something related for the metric computation itself, please restart the CI to see more details

File: ignite/handlers/checkpoint.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/handlers/checkpoint.py:712]: It is functional, what happens here is that in a single test @ydcjeff saves 3 times the same file. Normally, we do not overwrite existing files. With this change we will overwrite without any issue...

**Comment by vfdev-5** [comment]
> @vfdev-5 i don't think it's something related for the metric computation itself, please restart the CI to see more details

At first, you can try to checkout this PR locally and try to repro the issue if you have a Win32 machine. I think it is related as with new tests we have altered random state and it generates a corner case samples...

**Comment by KickItLikeShika** [comment]
@vfdev-5 i'm using Linux, i will try to investigate more and see what's the problem exactly

**Comment by sdesrozis** [comment]
@fco-dv could you help to track this issue on windows ? Thanks !

File: tests/ignite/contrib/metrics/test_gpu_info.py [hunk 3]
```diff
@@ -27,10 +27,11 @@ def no_site_packages():
 
 
 @pytest.mark.skipif(python_below_36, reason="No pynvml for python < 3.6")
-def test_no_pynvml_package(no_site_packages):
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_gpu_info.py:30]: As we are here, we can also remove `python_below_36` as we are now supporting >= 3.6 and droped 3.5

File: tests/ignite/contrib/metrics/test_gpu_info.py [hunk 4]
```diff
@@ -27,10 +27,11 @@ def no_site_packages():
...
+def test_no_pynvml_package():
 
-    with pytest.raises(RuntimeError, match="This contrib module requires pynvml to be installed."):
-        GpuInfo()
+    with patch.dict("sys.modules", {"pynvml.smi": None}):
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_gpu_info.py:32]: Do you have an idea why no_site_packages is not working anymore now ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update @ydcjeff

File: tests/ignite/contrib/metrics/test_gpu_info.py (hunk 4 above)
Comment (ydcjeff) [reply tests/ignite/contrib/metrics/test_gpu_info.py:32]: Could it be [we are importing `nvidia_smi` from pynvml](https://github.com/pytorch/ignite/blob/7480fed2d0389acb80b2f922e7242aeb091f8acf/ignite/contrib/metrics/gpu_info.py#L39) instead of `import pynvml` ?

**Comment by fco-dv** [comment]
I've checkout the PR , those tests are PASSED on my side:

> @fco-dv could you help to track this issue on windows ? Thanks !
````
(ignite-dev) D:\git\ignite>pytest  tests/ignite/contrib/metrics/test_average_precision.py -vvv
================================================= test session starts =================================================
platform win32 -- Python 3.7.7, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- d:\soft\miniconda3\envs\ignite-dev\python.exe
cachedir: .pytest_cache
rootdir: D:\git\ignite, inifile: setup.cfg
plugins: cov-2.8.1, forked-1.1.3, xdist-1.31.0
collected 15 items

tests/ignite/contrib/metrics/test_average_precision.py::test_no_sklearn PASSED                                   [  6%]
tests/ignite/contrib/metrics/test_average_precision.py::test_no_update PASSED                                    [ 13%]
tests/ignite/contrib/metrics/test_average_precision.py::test_input_types PASSED                                  [ 20%]
tests/ignite/contrib/metrics/test_average_precision.py::test_check_shape PASSED                                  [ 26%]
tests/ignite/contrib/metrics/test_average_precision.py::test_binary_input_N PASSED                               [ 33%]
tests/ignite/contrib/metrics/test_average_precision.py::test_multilabel_input_N PASSED                           [ 40%]
tests/ignite/contrib/metrics/test_average_precision.py::test_integration_binary_input_with_output_transform PASSED [ 46%]
tests/ignite/contrib/metrics/test_average_precision.py::test_integration_multilabel_input_with_output_transform PASSED [ 53%]
tests/ignite/contrib/metrics/test_average_precision.py::test_distrib_gpu SKIPPED                                 [ 60%]
tests/ignite/contrib/metrics/test_average_precision.py::test_distrib_cpu SKIPPED                                 [ 66%]
tests/ignite/contrib/metrics/test_average_precision.py::test_distrib_hvd SKIPPED                                 [ 73%]
tests/ignite/contrib/metrics/test_average_precision.py::test_multinode_distrib_cpu SKIPPED                       [ 80%]
tests/ignite/contrib/metrics/test_average_precision.py::test_multinode_distrib_gpu SKIPPED                       [ 86%]
tests/ignite/contrib/metrics/test_average_precision.py::test_distrib_single_device_xla SKIPPED                   [ 93%]
tests/ignite/contrib/metrics/test_average_precision.py::test_distrib_xla_nprocs SKIPPED                          [100%]

============================================ 8 passed, 7 skipped in 1.87s =============================================

````

**Comment by vfdev-5** [comment]
@fco-dv try complete test suite, maybe random seed is not the same between reduced and total runs.

**Comment by fco-dv** [comment]
@vfdev-5  I've launched the tests with `pytest tests -vvv `  and it passed , however I'm not sure to reproduce the command launched on the CI

```tests/ignite/contrib/metrics/test_average_precision.py::test_no_sklearn PASSED                                   [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_no_update PASSED                                    [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_input_types PASSED                                  [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_check_shape PASSED                                  [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_binary_input_N PASSED                               [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_multilabel_input_N PASSED                           [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_integration_binary_input_with_output_transform PASSED [ 22%]
tests/ignite/contrib/metrics/test_average_precision.py::test_integration_multilabel_input_with_output_transform PASSED [ 22%]
```

**Comment by sdesrozis** [comment]
@fco-dv Sorry for that but could you test reproducing the CI, maybe in a docker ? IMO It would be useful to have this dockerfile in the repo to ease the tracking of CI errors on windows (I suppose).

**Comment by vfdev-5** [comment]
@KickItLikeShika could you please inspect this failure on windows with `test_average_precision.py`. Thanks

**Comment by ydcjeff** [comment]
@vfdev-5 @KickItLikeShika Could this be failing case?
```py
>>> y_pred = torch.randint(0, 2, size=(10,)).long()
>>> n_iters = 16
>>> idx
13
>>> y_pred[idx : idx + batch_size]
tensor([], dtype=torch.int64)
```

**Comment by vfdev-5** [comment]
Thanks a lot @ydcjeff ! @KickItLikeShika can you make a pass over all contrib tests and fix all these issues we already talked about in some of your previous PRs.

**Comment by KickItLikeShika** [comment]
Thanks a lot @ydcjeff !
>  @KickItLikeShika can you make a pass over all contrib tests and fix all these issues we already talked about in some of your previous PRs.


@vfdev-5 I will do that

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot @ydcjeff  !
